src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -24,15 +35,60 @@ public synchronized Product getProduct(@Nullable Long id) throws ProductNotFound
         return products.stream()
                 .filter(product -> id.equals(product.getId()))
                 .findAny()
-                .orElseThrow(ProductNotFoundException::new);
+                .orElseThrow(() -> new ProductNotFoundException(id));
     }
 
     @Override
-    public synchronized List<Product> findProducts() {
-        return products.stream()
+    public synchronized List<Product> findProducts(String query, SortField sortField, SortOrder sortOrder) {
+        List<Product> processingProducts = products.stream()
                 .filter(product -> product.getPrice() != null)
                 .filter(product -> product.getStock() > 0)
                 .collect(Collectors.toList());
+
+        if (query != null && !query.isEmpty()) {
+            processingProducts = getQueriedProducts(processingProducts, query);
+        }
+
+        if (sortField != null && sortOrder != null) {
+            processingProducts = getSortedProducts(processingProducts, sortField, sortOrder);
+        }
+
+        return processingProducts;
+    }
+
+    private List<Product> getSortedProducts(List<Product> products, SortField sortField, SortOrder sortOrder) {
+        Comparator<Product> fieldComparator = (sortField == SortField.description)
+                ? Comparator.comparing(Product::getDescription)
+                : Comparator.comparing(Product::getPrice);
+        if (sortOrder == SortOrder.desc) {
+            fieldComparator = fieldComparator.reversed();
+        }
+
+        return products.stream()
+                .sorted(fieldComparator)
+                .collect(Collectors.toList());
+    }
+
+    private List<Product> getQueriedProducts(List<Product> products, String query) {
+        String[] queryWords = splitToWords(query);
+
+        return products.stream()
+                .filter(product -> isContainingQueryWords(product.getDescription(), queryWords))
+                .sorted(Comparator.comparing(product -> getQueryRelevance(((Product) product).getDescription(), queryWords)).reversed())
```

Comment:
So you traverse products twice here, first when filter, and second when sort. Check what you can do to traverse only once.